src/middlewared/middlewared/plugins/ldap.py

```diff
@@ -0,0 +1,639 @@
+
+        if not await self.nslcd_status():
+            await self.nslcd_cmd('onestart')
+        else:
+            await self.nslcd_cmd('onerestart')
+
+        if has_samba_schema:
+            await self.middleware.call('etc.generate', 'smb')
+            await self.middleware.call('service.restart', 'smb')
+
+        await self.middleware.call('ldap.fill_ldap_cache')
+
+    @private
+    async def stop(self):
+        ldap = await self.config()
+        await self.middleware.call('datastore.update', self._config.datastore, ldap['id'], {'ldap_enable': False})
+        await self.__set_state(DSStatus['LEAVING'])
+        await self.middleware.call('etc.generate', 'rc')
+        await self.middleware.call('etc.generate', 'nss')
+        await self.middleware.call('etc.generate', 'ldap')
+        await self.middleware.call('etc.generate', 'pam')
+        if ldap['has_samba_schema']:
+            await self.middleware.call('etc.generate', 'smb')
+            await self.middleware.call('service.restart', 'smb')
+        await self.middleware.call('cache.pop', 'LDAP_State')
+        await self.nslcd_cmd('onestop')
+
+    @private
+    @job(lock='fill_ldap_cache')
+    def fill_ldap_cache(self, force=False):
+        """
+        """
+        if self.middleware.call_sync('cache.has_key', 'LDAP_cache') and not force:
+            raise CallError('LDAP cache already exists. Refusing to generate cache.')
+
+        self.middleware.call_sync('cache.pop', 'LDAP_cache')
+        pwd_list = pwd.getpwall()
+        grp_list = grp.getgrall()
+
+        local_uid_list = list(u['uid'] for u in self.middleware.call_sync('user.query'))
+        local_gid_list = list(g['gid'] for g in self.middleware.call_sync('group.query'))
+        cache_data = {'users': [], 'groups': []}
+
+        for u in pwd_list:
+            is_local_user = True if u.pw_uid in local_uid_list else False
+            if is_local_user:
+                continue
+
+            cache_data['users'].append({
+                'name': u.pw_name,
+                'uid': u.pw_name,
+                'uidNumber': u.pw_uid,
+                'gidNumber': u.pw_gid,
+                'gecos': u.pw_gecos,
+                'homeDirectory': u.pw_dir,
+                'loginShell': u.pw_shell,
+            })
+
+        for g in grp_list:
+            is_local_user = True if g.gr_gid in local_gid_list else False
+            if is_local_user:
+                continue
+
+            cache_data['groups'].append({
+                'name': g.gr_name,
+                'group': g.gr_name,
+                'gidNumber': g.gr_gid,
+                'members': g.gr_mem,
+            })
+
+        self.middleware.call_sync('cache.put', 'LDAP_cache', cache_data, 86400)
+
+    @private
+    async def get_ldap_cache(self):
+        """
+        """
+        if not await self.middleware.call('cache.has_key', 'LDAP_cache'):
+            await self.middleware.run_in_thread(self.fill_ldap_cache)
```

Comment:
await self.middleware.call('ldap.fill_ldap_cache') is supposed to return a job
which we can then do:
await job.wait()